Allow authentication mechanism selection for a REST endpoint with annotation (#36504)
quarkusio/quarkus: docs/src/main/asciidoc/security-authentication-mechanisms.adoc

Ensure that the value of the `auth-mechanism` property matches the authentication scheme supported by `HttpAuthenticationMechanism`, for example, `basic`, `bearer`, or `form`.

=== Endpoint-specific authentication mechanisms

It is possible to select authentication mechanism specific for each Jakarta REST endpoint and Jakarta REST resource with annotation.
This feature is only enabled when <<proactive-auth>> is disabled, for authentication selection must happen after a REST endpoint has been matched.
Here is how you can select <<openid-connect-authentication>> mechanism per a REST endpoint basis:

[source,properties]
----

> quarkus.http.auth.proactive=false

Reviewer: Michal, what I'm not quite sure, now that `@Bearer`, and therefore `@HttpAuthMechanism("bearer")` is treated like `@Authenticated`, why it can't work with the proactive mode enabled. We know that `@Bearer` means the authentication is required.

Lets say we only have `quarkus-oidc`, a single service tenant, and `@Bearer` - why we'd not make it working with the proactive mode when the proactive mode works with `@Authenticated` instead of `@Bearer` in this case ?

Author: Because RESTEasy Reactive (I know little less how Classic works, but I assume it's pretty similar) needs to select correct endpoint based on HTTP request - it's not just path. We need to let RESTEasy Reactive do it and we can't pre-process it.

=>

The RESTEasy Reactive needs to run before authentication happens, and proactive authentication happens before the RESTEasy Reactive starts processing. Moreover if we want proactive auth happen after RESTEasy Reactive started processing, it couldn't be based on `RoutingContext` (what I mean: no `event.next()` or `event.fail()` etc.) it would case whole lot of problems. The RR logic is just too complext, we can't abstract it and create some `RoutingContext` to endpoint mapper.

Anyway, you probably mentioned, but let me emphasize I added insurance - if authentication is invoked before the request is matched with endpoint, authentication fails. That together with `Bearer` == `@Authenticated` should mean we still follow security first.

Author: Ah, you can ignore my last paragraph, now I mentioned you commented on it above :-)

Reviewer: Hi Michal, @michalvavrik, I have to admit I did not quite get the message, why all of it is not a problem for `@Authenticated`  but is a problem for say `@Basic` which is now effectively `@Authenticated` but only a mechanism restriction. Or did you mean with your last comment it would for now for either mode (proactive on/off) ?

Author: > Hi Michal, @michalvavrik, I have to admit I did not quite get the message, why all of it is not a problem for @Authenticated but is a problem for say @Basic which is now effectively @Authenticated but only a mechanism restriction. Or did you mean with your last comment it would for now for either mode (proactive on/off) ?

It is not problem for `@Authenticated` because in case before `@Authenticated` security interceptor is applied there was also HTTP Security Policy, the very same identity is used (the one that is already resolved, right?). But security identity can only be resolved when you have selected authentication mechanism.

Reviewer: @michalvavrik Hi,

>It is not problem for @Authenticated because in case before @Authenticated security interceptor is applied there was also HTTP Security Policy

Lets take a case when neither `@Authenticated` nor `@Basic` have any security policy configurations. Can you please clarify why `@Authenticated` will work with proactive authentication but `@Basic` not ?

Thanks

Author: > Can you please clarify why @Authenticated will work with proactive authentication but @Basic not ?

@sberyozkin 

I can. Proactive authentication gets identity, but accepts anonymous one. When request arrives to `AuthenticatedInterceptor` (-> `AuthenticatedCheck`) it doesn't really perform authentication as identity has been resolved . This is what happens in meta code:

is identity anonymous? UnauthorizedException
is identity not anonymous? nothing

Please did you mention there is no call to identity provider or authentication mechanism? Because that only happens when proactive authentication is disabled.

How does it differ to `@Basic`? The `@Basic` enforces `auth-mechanism=basic`, which means no other mechanism may authenticate (it is 100 % same as when specified mech via path matching code). But when proactive auth happens, mechanism is not yeth selected. Of course `@Authenticated` itself would work, just applied on the wrong mechanism.

Author: hello @sberyozkin ,

let me answer your question about proactive authentication from scratch.

Proactive authentication means that HTTP request is authenticated from Vert.x route handler way before you know anything about REST stack. If you authenticate before you know the REST stack, you don't know anything about annotations like `@Bearer`, because they are on the endpoints.

Therefore, if proactive authentication were enabled, it would happen before `@Bearer` was recognized and when you got down to the RESTEasy Reactive, you would authenticate for a second time.

I don't believe this can be blocker, because I remember annotation `@Tenant` and RESTEasy exception mappers. There is no requirement for every single feature to work with proactive authentication.

Ensure that the value of the `auth-mechanism` property matches the authentication scheme supported by `HttpAuthenticationMechanism`, for example, `basic`, `bearer`, or `form`.

> === Endpoint-specific authentication mechanisms

Author: I am fine with propose title and I'll change it, but let me provide some context.

> This may read strange but the endpoint specific, path based authentication is already supported, it is really about using the annotations vs HTTP security policy to enable this feature.

Yes, but the annotation-based approach is more powerful because it is capable to select endpoint, which the HTTP Security Policy never will. You can have several endpoints with a same path and HTTP method but they will differ by content type etc. just FYI

> === Endpoint-specific authentication mechanisms

Reviewer: @michalvavrik I'm not disagreeing, I'm just saying that we already allow to do it with HTTP Security Policy, so when users will see one section "Path based authentication" where we say how to use HTTP Security Policy and `Endpoint specific authentication` where the annotations are used, they will be confused IMHO. So the suggestion is to make it clear that one section is about using the policy, another - annotations, to achieve the same goal

> === Endpoint-specific authentication mechanisms

Reviewer: @michalvavrik Ma be we can have a section `Path-based authentication` then the annotation based selection for JAX-RS goes next and we say we recommend it, the existing one for HTTP security policy goes last ? May be also clarify it as `Use annotations to enable path-based authentication for JAX-RS endpoints`

> It is possible to select authentication mechanism specific for each Jakarta REST endpoint and Jakarta REST resource with annotation.

Reviewer: ```suggestion
It is possible to use annotations to select an authentication mechanism specific to each Jakarta REST endpoint and resource.
```

> This feature is only enabled when <<proactive-auth>> is disabled, for authentication selection must happen after a REST endpoint has been matched.

Reviewer: ```suggestion
This feature is only enabled when <<proactive-auth>> is disabled due to the fact that the annotations can only be used to select authentication mechanisms after a REST endpoint has been matched.
```

Here is how you can select <<openid-connect-authentication>> mechanism per a REST endpoint basis:

[source,properties]
----
quarkus.http.auth.proactive=false

> # combine HTTP permissions with the endpoint-specific authentication mechanism

Reviewer: IMHO we should not start with giving an HTTP security example when introducing this feature, I'd add a subsection like `==== How to combine it with HTTP security policy` and give this `applies-to` example there.
In the introduction itself, I propose to give `BasicAuthentication` and `AuthorizationCodeFlow` as users did ask for this kind of set up before

Author: Sure, combination with the HTTP Security Policies is rather for compatibility than recommended use case. I'll move it.
